README.md
<p align="center" style='font-style: italic;'>
The format is based on [Best-README-Template](https://github.com/othneildrew/Best-README-Template/blob/master/BLANK_README.md).
</p>

<!-- PROJECT LOGO -->
<br />
<p align="center">
  <a href="https://github.com/DannySauval/Flappy_Bird_AI">
    <img src="assets/bluebird-midflap.png" alt="Logo" width="80">
  </a>

  <h3 align="center">Flappy Bird AI</h3>

  <p align="center">
    Flappy Bird game to play around with ANN
    <br />
    <a href="https://github.com/DannySauval/Flappy_Bird_AI"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://github.com/DannySauval/Flappy_Bird_AI">View Demo</a>
    ·
    <a href="https://github.com/DannySauval/Flappy_Bird_AI/issues">Report Bug</a>
    ·
    <a href="https://github.com/DannySauval/Flappy_Bird_AI/issues">Request Feature</a>
  </p>
</p>



<!-- TABLE OF CONTENTS -->
<details open="open">
  <summary><h2 style="display: inline-block">Table of Contents</h2></summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgements">Acknowledgements</a></li>
  </ol>
</details>



<!-- ABOUT THE PROJECT -->
## About The Project

I wanted to do make a simple game using what I learned about AI lastly. The idea was to be as simple as possible, but start from scratch. I forbid myself to use anything else than numpy and pygame or standard libraries.
I decided to go with Flappy Bird which is a simple game to implement qnd to play. Only a few data are required for the bird to play, and the output decidion is easy, either you jump or you don't.

### Built With

* []()
* []()
* []()



<!-- GETTING STARTED -->
## Getting Started

To get a local copy up and running follow these simple steps.

### Prerequisites

There are a few requirements to be able to run Flappy Bird AI.
* Python 3.x : <a hfref="https://www.python.org/downloads/">https://www.python.org/downloads/</a>
* pygame
  ```sh
  pip install pygame
  ```
* numpy
  ```sh
  pip install numpy
  ```

### Installation

1. Clone the repo
   ```sh
   git clone https://github.com/DannySauval/Flappy_Bird_AI.git
   ```
2. Run FlappyBird.py
   ```sh
   py.exe FlappyBird.py
   ```

<!-- USAGE EXAMPLES -->
## Usage

Use this space to show useful examples of how a project can be used. Additional screenshots, code examples and demos work well in this space. You may also link to more resources.

_For more examples, please refer to the [Documentation](https://example.com)_


<!-- LICENSE -->
<!--## License

Distributed under the MIT License. See `LICENSE` for more information.-->



<!-- CONTACT -->
## Contact

Danny SAUVAL 
- Twitter : [@DSauval](https://twitter.com/DSauval)
- Youtube : [DannySauval](https://www.youtube.com/channel/UC_pCdmt0cSF49st1b5q6yIQ)
- Email : sauvaldanny@gmail.com

Project Link: [https://github.com/DannySauval/Flappy_Bird_AI](https://github.com/DannySauval/Flappy_Bird_AI)

<!-- ACKNOWLEDGEMENTS -->
## Acknowledgements

* []()
* []()
* []()


CHANGELOG.md
# # Flappy_Bird_AI CHangelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]

## 2020-12-10 : Commit : "Optimization. Commenting."
### Changed

1. Finished commenting.
2. More efficient when going through the birds list.
3. Slider to change framerate in real time.
4. I finished all what I planned.

## [Unreleased]

## 2020-12-10 : Commit : "Game in a class. Fixed pipe system."
### Changed

1. Overall structure :
	- Now the game is in a class.
	- A lot of simplifying.
2. Pipes system : 
	- Fixed a bug, when resetting a pipe's position, I had forgotten to call for new random height.
	- Fixed the bug where the space between pipes was not right. Not spacing is a parameter.
3. NeuralNetwork :
	- Calibrating itself using NeuralNetwork.init() after importing the class.

## [Unreleased]

## 2020-12-09 : Commit : "New pipe system. Cleaned Bird Class."
### Changed

1. Pipes system :
	- New optimized pipe system. Only uses 3 pipes.
	- No more bug when a pipe's position overflow.

2. Bird :
	- Bird class is cleaned.
	- Loading Bird image only once.

### To do next
1. Create FlappyBird class, which wille serve as game handler.
2. Remove weights files from GitHub and git ignore pycache.

Bird.py
import random
import pygame

from NeuralNetwork import NeuralNetwork

from settings import BIRD, WORLD

class Bird:
    high_score = 0
    alive_birds = 0

    bird_surface : pygame.Surface = None

    def __init__(self, weights_list):
        self.score = 0   # Number of pipes
        self.fitness = 0 # Number of frames until dead
        self.bird_movement = 0
        self.alive = True
        self.next_pipe_index = 0
        self.net = NeuralNetwork(weights_list)

        self.gravity = BIRD['GRAVITY']
        self.jump_height = BIRD['JUMP_HEIGHT']
        # If it's the first bird we declare, we want to load the image
        self.bird_rect = Bird.bird_surface.get_rect(center = (50, WORLD['HEIGHT']/2))

        Bird.alive_birds += 1

    @staticmethod
    def clsinit():
        """
            Initialize the class' bird_surface. We do it here so that we do it only once.
        """
        Bird.bird_surface = pygame.image.load('assets/bluebird-midflap.png').convert_alpha()

    def update_highscore(self):
        """
            Updates class' high_score value.
        """
        if self.score > Bird.high_score:
            Bird.high_score = int(self.score)

    def check_collision(self, pipes):
        """
            Returns False if there is a collision or if the bird gets out of the screen.
            Returns True otherwise.
        """
        for pipe in pipes:
            if self.bird_rect.colliderect(pipe):
                return False
        if self.bird_rect.top <= -90 or self.bird_rect.bottom >= WORLD['FLOOR_BIAS']:
            return False
        return True

    def make_decision(self, pipes):
        if self.alive:
            # if len(pipes) > self.next_pipe_index + 1:
            if pipes[self.next_pipe_index].right - self.bird_rect.left < 0:
                self.next_pipe_index = (self.next_pipe_index+2)%6 # Next bottom pipe
                self.score += 1
            inputs = []
            # X Distance to next pipes
            inputs.append(pipes[self.next_pipe_index].left - self.bird_rect.right)
            # Y Distance to top pipe
            inputs.append(self.bird_rect.top - pipes[self.next_pipe_index+1].bottom)
            # Y Distance to bottom pipe
            inputs.append(pipes[self.next_pipe_index].top - self.bird_rect.bottom)
            # Y Position of the bird
            inputs.append(self.bird_rect.centery)

            if self.net.make_decision(inputs):
                self.jump()

    def jump(self):
        self.bird_movement = -self.jump_height

    def update(self, screen, pipe_list):
        if self.alive:
            self.fitness += BIRD['FITNESS_STEP']
            self.update_highscore()

            self.bird_movement += self.gravity
            self.bird_rect.centery += self.bird_movement
            screen.blit(Bird.bird_surface, self.bird_rect)

            self.alive = self.check_collision(pipe_list)
            if not self.alive:
                Bird.alive_birds -= 1
       

Slider.py
"""
    Slider class from https://www.dreamincode.net/forums/topic/401541-buttons-and-sliders-in-pygame/
"""

import pygame

WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
RED = (255, 50, 50)
YELLOW = (255, 255, 0)
GREEN = (0, 255, 50)
BLUE = (50, 50, 255)
GREY = (200, 200, 200)
ORANGE = (200, 100, 50)
CYAN = (0, 255, 255)
MAGENTA = (255, 0, 255)
TRANS = (1, 1, 1)

from settings import WORLD

class Slider():
    def __init__(self, name, val, mini, maxi, pos, font):
        self.val = val  # start value
        self.maxi = maxi  # maximum at slider position right
        self.mini = mini  # minimum at slider position left
        self.xpos = pos  # x-location on screen
        self.ypos = WORLD['HEIGHT']
        self.surf = pygame.surface.Surface((100, 50))
        self.hit = False  # the hit attribute indicates slider movement due to mouse interaction
        self.name = name
        self.font = pygame.font.SysFont("Arial", 16)

        self.txt_surf = self.font.render(self.name + " " + str(self.val), 1, BLACK)
        self.txt_rect = self.txt_surf.get_rect(center=(50, 15))

        # Static graphics - slider background #
        self.surf.fill(GREY)
        pygame.draw.rect(self.surf, RED, [0, 0, 100, 50], 3)
        pygame.draw.rect(self.surf, WHITE, [10, 30, 80, 5], 0)

        # dynamic graphics - button surface #
        self.button_surf = pygame.surface.Surface((20, 20))
        self.button_surf.fill(TRANS)
        self.button_surf.set_colorkey(TRANS)
        pygame.draw.circle(self.button_surf, BLACK, (10, 10), 6, 0)
        pygame.draw.circle(self.button_surf, ORANGE, (10, 10), 4, 0)

    def draw(self, screen):
        """ Combination of static and dynamic graphics in a copy of
    the basic slide surface
    """
        # static
        surf = self.surf.copy()

        # dynamic
        surf.fill(GREY)
        pygame.draw.rect(surf, RED, [0, 0, 100, 50], 3)
        pygame.draw.rect(surf, WHITE, [10, 30, 80, 5], 0)
        surf.blit(self.txt_surf, self.txt_rect)
        pos = (10+int((self.val-self.mini)/(self.maxi-self.mini)*80), 33)
        self.button_rect = self.button_surf.get_rect(center=pos)
        surf.blit(self.button_surf, self.button_rect)
        self.button_rect.move_ip(self.xpos, self.ypos)  # move of button box to correct screen position

        # screen
        screen.blit(surf, (self.xpos, self.ypos))

    def move(self):
        """
    The dynamic part; reacts to movement of the slider button.
    """
        self.val = int((pygame.mouse.get_pos()[0] - self.xpos - 10) / 80 * (self.maxi - self.mini) + self.mini)
        if self.val < self.mini:
            self.val = self.mini
        if self.val > self.maxi:
            self.val = self.maxi
        self.txt_surf = self.font.render(self.name + " " + str(self.val), 1, BLACK)

mainYT.py
import pygame, sys, random

def draw_floor():
    screen.blit(floor_surface, (floor_x_pos, FLOOR_BIAS))
    screen.blit(floor_surface, (floor_x_pos + WIDTH, FLOOR_BIAS))

def create_pipe():
    random_pipe_pos = random.choice(pipe_height)
    bottom_pipe = pipe_surface.get_rect(midtop = (350, random_pipe_pos))
    top_pipe = pipe_surface.get_rect(midbottom = (350, random_pipe_pos - 150))
    return bottom_pipe, top_pipe

def move_pipes(pipes):
    for pipe in pipes:
        pipe.centerx -= 2.5
    return pipes

def draw_pipes(pipes):
    for pipe in pipes:
        if pipe.bottom >= 512:
            screen.blit(pipe_surface, pipe)
        else:
            flip_pipe = pygame.transform.flip(pipe_surface, False, True)
            screen.blit(flip_pipe, pipe)

def rotate_bird(bird):
    new_bird = pygame.transform.rotozoom(bird, -5*bird_movement, 1)
    return new_bird

def check_collision(pipes):
    for pipe in pipes:
        if bird_rect.colliderect(pipe):
            death_sound.play()
            return False
    if bird_rect.top <= -90 or bird_rect.bottom >= FLOOR_BIAS:
        return False
    return True

def bird_animation():
    new_bird = bird_frames[bird_index]
    new_bird_rect = new_bird.get_rect(center = (50, bird_rect.centery))
    return new_bird, new_bird_rect

def score_display(game_state):
    if game_state == 'main_game':
        score_surface = game_font.render(f'Score : {int(score)}', True, (255, 255, 255))
        score_rect = score_surface.get_rect(center = (WIDTH/2, 50))
        screen.blit(score_surface, score_rect)
    if game_state == 'game_over':
        score_surface = game_font.render(f'Score : {int(score)}', True, (255, 255, 255))
        score_rect = score_surface.get_rect(center = (WIDTH/2, 50))
        screen.blit(score_surface, score_rect)

        high_score_surface = game_font.render(f'Score : {int(high_score)}', True, (255, 255, 255))
        high_score_rect = high_score_surface.get_rect(center = (WIDTH/2, HEIGHT-90))
        screen.blit(high_score_surface, high_score_rect)

def update_score(score, high_score):
    if score > high_score:
        high_score = score
    return high_score

pygame.mixer.pre_init(frequency = 44100, size = 16, channels = 1, buffer = 512)
pygame.init()

# CONSTANTS
WIDTH = 288
FLOOR_BIAS = 450
HEIGHT = 512
JUMP_HEIGHT = 4
PIPE_SPEED = 2.5
PIPE_SPAWN_SPEED = 1200
SCORE_INC = 0.05
gravity = 0.125
bird_movement = 0
game_active = True
score = 0
high_score = 0


# Window setup
screen = pygame.display.set_mode((WIDTH, HEIGHT))
icon = pygame.image.load('assets/flappy_bird_icon.png')
pygame.display.set_caption("Flappy Bird AI")
pygame.display.set_icon(icon)

# Setting up framerate
clock = pygame.time.Clock()

# Font
game_font = pygame.font.Font('assets/04B_19.ttf', 30)

# Images import
bg_surface = pygame.image.load('assets/background-day.png').convert()

floor_surface = pygame.image.load('assets/base.png').convert()
floor_x_pos = 0

# bird_surface = pygame.image.load('assets/bluebird-midflap.png').convert_alpha()
# bird_rect = bird_surface.get_rect(center = (50, HEIGHT/2))

bird_downflap = pygame.image.load('assets/bluebird-downflap.png').convert_alpha()
bird_midflap = pygame.image.load('assets/bluebird-midflap.png').convert_alpha()
bird_upflap = pygame.image.load('assets/bluebird-upflap.png').convert_alpha()
bird_frames = [bird_downflap, bird_midflap, bird_upflap]
bird_index = 0
bird_surface = bird_frames[bird_index]
bird_rect = bird_surface.get_rect(center = (50, HEIGHT/2))
BIRDFLAP = pygame.USEREVENT + 1
pygame.time.set_timer(BIRDFLAP, 200)

pipe_surface = pygame.image.load('assets/pipe-green.png')
pipe_list = []
SPAWNPIPE = pygame.USEREVENT
pygame.time.set_timer(SPAWNPIPE, PIPE_SPAWN_SPEED)
pipe_height = [200, 300, 400]

game_over_surface = pygame.image.load('assets/message.png').convert_alpha()
game_over_rect = game_over_surface.get_rect(center = (WIDTH/2, HEIGHT/2))

# Sound
flap_sound = pygame.mixer.Sound('sound/sfx_wing.wav')
death_sound = pygame.mixer.Sound('sound/sfx_hit.wav')

# Main loop
while True:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit()
            sys.exit()
        if event.type == pygame.KEYDOWN:
            if event.key == pygame.K_SPACE and game_active:
                bird_movement = -JUMP_HEIGHT
                flap_sound.play()
            if event.key == pygame.K_SPACE and game_active == False:
                game_active = True
                pipe_list.clear()
                bird_rect.center = (50, 256)
                bird_movement = 0
                score = 0


        if event.type == SPAWNPIPE:
            pipe_list.extend(create_pipe())
        if event.type == BIRDFLAP:
            if bird_index < 2:
                bird_index += 1
            else:
                bird_index = 0
            bird_surface, bird_rect = bird_animation()

    screen.blit(bg_surface, (0, 0))

    if game_active:
        # Bird
        bird_movement += gravity
        rotated_bird = rotate_bird(bird_surface)
        bird_rect.centery += bird_movement
        screen.blit(rotated_bird, bird_rect)
        game_active = check_collision(pipe_list) 

        # Pipes 
        pipe_list = move_pipes(pipe_list)
        draw_pipes(pipe_list)
        score += SCORE_INC
        score_display('main_game')
    else:
        screen.blit(game_over_surface, game_over_rect)
        high_score = update_score(score, high_score)
        score_display('game_over')

    # Floor
    floor_x_pos -= 1
    draw_floor()
    
    if floor_x_pos <= -WIDTH:
        floor_x_pos = 0

    pygame.display.update()
    clock.tick(120)

pygame.quit()


NeuralNetwork.py
import numpy as np
import random

class NeuralNetwork:
    """
        Class to generate a NeuralNetwork.
        One hidden layer.
    """
    calibration = 0.58

    def __init__(self, weights_list):
        self.weights = weights_list
        self.output = 0

    def feedforward(self, x_input):
        """
            Process the feedforward of the NeuralNetwork.
        """
        layer1 = self.sigmoid(np.dot(x_input, self.weights[0]))
        layer2 = self.sigmoid(np.dot(layer1, self.weights[1]))
        # (0.99896684/2) is substracted from layer2 to get a probability of 0.5.
        # I am not really sure about this part and I need to dig a little bit more.
        # self.output = self.sigmoid(layer2-(0.99896684/2))
        self.output = self.sigmoid(layer2-self.calibration)

    def make_decision(self, x_input):
        """
            Make a decision based on the output of the Neural Network.
        """
        self.feedforward(x_input)
        if self.output > 0.5:
            return True
        return False

    def __str__(self):
        return f'{self.output}'

    @staticmethod
    def clsinit():
        mean = 0
        N = 1000

        for i in range(0, N):
            x = np.random.randn(1,4)
            weights = [np.random.rand(4, 7), np.random.rand(7, 1)]
            nn = NeuralNetwork(weights)
            nn.make_decision(x)
            mean += nn.output

        NeuralNetwork.calibration = mean / N

    @staticmethod
    def sigmoid(x_input):
        """
            Sigmoid activation function.
            Mathematical definition of the sigmoid function.
        """
        return 1/(1+np.exp(-x_input))

if __name__ == '__main__':
    mean = 0
    N = 1000

    for i in range(0, N):
        x = np.random.randn(1,4)
        weights = [np.random.rand(4, 7), np.random.rand(7, 1)]
        nn = NeuralNetwork(weights)
        nn.make_decision(x)
        mean += nn.output

    mean /= N

    print(mean)


FlappyBird.py
import numpy as np
import sys
import random
import pygame

from settings import WORLD, PIPE, BIRD, PLAYER
from Bird import Bird
from NeuralNetwork import NeuralNetwork

from Slider import Slider

class FlappBird:
    def __init__(self):
        pygame.init()

        # Window setup
        self.screen = pygame.display.set_mode((WORLD['WIDTH'], WORLD['HEIGHT']+100))
        pygame.display.set_caption("Flappy Bird AI")
        pygame.display.set_icon(pygame.image.load('assets/flappy_bird_icon.png'))

        Bird.clsinit() # Initialization of Bird and NeuralNetwork classes
        NeuralNetwork.clsinit()

        self.game_active = True

        # Setting up framerate
        self.clock = pygame.time.Clock()

        # Font & Images import
        self.game_font = pygame.font.Font('assets/04B_19.ttf', 20)
        self.bg_surface = pygame.image.load('assets/background-day.png').convert()
        self.floor_surface = pygame.image.load('assets/base.png').convert()
        self.floor_x_pos = 0

        self.pipe_surface = pygame.image.load('assets/pipe-green.png')
        self.pipe_list = []

        for i in range(3):
            self.pipe_list.extend(self.create_pipe(i))

        self.game_over_surface = pygame.image.load('assets/message.png').convert_alpha()
        self.game_over_rect = self.game_over_surface.get_rect(center = (WORLD['WIDTH']/2, WORLD['HEIGHT']/2))

        self.sliders_bg = pygame.surface.Surface((WORLD['WIDTH'], 100))
        self.sliders_bg.fill((0, 0, 0))

        self.birds = self.generate_pop(mutated=False)

        self.best_bird_weights = []
        self.best_bird_fitness = 0

        self.timer = 0
        self.offscreen_pipe_index = 0
        self.add_pipe_timer = 2*PIPE['PIPE_SPACING']/PIPE['PIPE_SPEED']

        self.current_generation = 0
        self.best_gen_highscore = 0

        self.sliders = [Slider("FPS", WORLD['FPS'], 10, 2000, 0, self.game_font)]

    def run(self):
        """
            Runs the main loop.
        """
        while True:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    pygame.quit()
                    sys.exit()
                if event.type == pygame.KEYDOWN:
                    if event.key == pygame.K_SPACE and self.game_active:
                        for bird in self.birds:
                            bird.jump()
                if event.type == pygame.MOUSEBUTTONDOWN:
                    pos = pygame.mouse.get_pos()
                    for s in self.sliders:
                        if s.button_rect.collidepoint(pos):
                            s.hit = True
                elif event.type == pygame.MOUSEBUTTONUP:
                    for s in self.sliders:
                        s.hit = False

            keys = pygame.key.get_pressed()  #checking pressed keys
            if keys[pygame.K_LEFT]:
                self.move_pipes()
                self.timer += 1

            if not PLAYER['MANUAL_PLAY']:
                # For each birds, decide if it's going to jump or not.
                for bird in self.birds:
                    bird.make_decision(self.pipe_list)

            # Recycle offscreen pipe every add_pipe_timer frames.
            self.timer += 1
            if self.timer >= self.add_pipe_timer:
                self.pipe_list[self.offscreen_pipe_index], self.pipe_list[self.offscreen_pipe_index+1] = self.create_pipe(1)

                self.offscreen_pipe_index = (self.offscreen_pipe_index+2)%6
                self.timer = 0
                self.add_pipe_timer = PIPE['PIPE_SPACING']/PIPE['PIPE_SPEED']

            # Background
            self.screen.blit(self.bg_surface, (0, 0))

            if self.game_active:
                # Bird
                if Bird.alive_birds == 0:
                    self.game_active = False
                for bird in self.birds:
                    bird.update(self.screen, self.pipe_list)

                # Pipes
                self.move_pipes()
                self.draw_pipes()
                self.data_display()
                
                if not self.game_active:
                    self.reset_game()
            else:
                self.screen.blit(self.game_over_surface, self.game_over_rect)

            # Floor
            self.draw_floor()

            # Move slides
            for slider in self.sliders:
                if slider.hit:
                    slider.move()

            self.screen.blit(self.sliders_bg, (0, WORLD['HEIGHT']))

            # Draws slider
            for slider in self.sliders:
                slider.draw(self.screen)

            pygame.display.update()
            self.clock.tick(self.sliders[0].val)

        pygame.quit()

    def reset_game(self):
        """
            Reset all variables to start again with a new bird batch.
        """
        # Get the bird with the best fitness.
        for bird in self.birds:
            if bird.fitness > self.best_bird_fitness:
                self.best_bird_weights = bird.net.weights
                self.best_bird_fitness = bird.fitness

        if Bird.high_score > self.best_gen_highscore:
            self.best_gen_highscore = Bird.high_score

        self.birds.clear()
        self.birds = self.generate_pop(mutated=True)
        Bird.high_score = 0
        self.game_active = True
        self.timer = 0

        self.offscreen_pipe_index = 0
        self.add_pipe_timer = 2*PIPE['PIPE_SPACING']/PIPE['PIPE_SPEED']
        
        self.reset_pipes()

        self.current_generation += 1

    def draw_floor(self):
        """
            Draws the floor and handles the landscape scrolling.
        """
        self.floor_x_pos -= 1
        self.screen.blit(self.floor_surface, (self.floor_x_pos, WORLD['FLOOR_BIAS']))
        self.screen.blit(self.floor_surface, (self.floor_x_pos + WORLD['WIDTH'], WORLD['FLOOR_BIAS']))

        if self.floor_x_pos <= -WORLD['WIDTH']:
            self.floor_x_pos = 0

    def create_pipe(self, pipe_number):
        """
            Returns a new pair of pipes, with random height.
        """
        random_pipe_pos = random.choice([200, 300, 400])
        bottom_pipe = self.pipe_surface.get_rect(midtop = ((pipe_number*PIPE['PIPE_SPACING'])+350, random_pipe_pos))
        top_pipe = self.pipe_surface.get_rect(midbottom = ((pipe_number*PIPE['PIPE_SPACING'])+350, random_pipe_pos - 150))

        return bottom_pipe, top_pipe

    def data_display(self):
        """
            Prints what we want to be displayed on the screen.
        """
        self.blit_text(f'GEN HS : {int(self.best_gen_highscore)}', (WORLD['WIDTH']-100, 10), (255, 255, 255))
        self.blit_text(f'HS : {int(Bird.high_score)}', (WORLD['WIDTH']-100, 30), (255, 255, 255))
        self.blit_text(f'BA : {int(Bird.alive_birds)}', (WORLD['WIDTH']-100, 50), (255, 255, 255))
        self.blit_text(f'GEN : {int(self.current_generation)}', (WORLD['WIDTH']-100, 70), (255, 255, 255))

    def blit_text(self, text, pos, color):
        """
            Blits test on the screen using specified text string, position and color.
        """
        score_surface = self.game_font.render(text, True, color)
        score_rect = score_surface.get_rect(center = pos)
        self.screen.blit(score_surface, score_rect)

    def move_pipes(self):
        """
            Shifts all the pipes to the left.
        """
        for pipe in self.pipe_list:
            pipe.centerx -= PIPE['PIPE_SPEED']

    def draw_pipes(self):
        """
            Draws the bottom&top pipes on the screen.
        """
        for pipe in self.pipe_list:
            if pipe.bottom >= 512:
                self.screen.blit(self.pipe_surface, pipe)
            else:
                flip_pipe = pygame.transform.flip(self.pipe_surface, False, True)
                self.screen.blit(flip_pipe, pipe)

    def reset_pipes(self):
        """
            Reset the pipes to initial position.
        """
        for i in range(3):
            self.pipe_list[i*2], self.pipe_list[(i*2)+1] = self.create_pipe(i)

    def model_mutate(self, weights_list):
        """
            Mutate the input weights.
            80% chance mutation. Mutation between [-0.5; 0.5[, uniform distribution.
        """
        mutated_weights = [weights_list[0].copy(), weights_list[1].copy()]

        for weights in mutated_weights:
            for weight in weights:
                for i, _ in enumerate(weight):
                    if np.random.uniform(0, 1) > 0.8:
                        change = np.random.uniform(-0.5, 0.5)
                        weight[i] += change
        return mutated_weights

    def generate_pop(self, mutated=False):
        """
            Generate population randomly, or from a bird that we evolve.
        """
        gen_birds = []

        for _ in range(BIRD['POP_SIZE']):
            if not mutated:
                gen_birds.append(Bird([np.random.randn(4, 7), np.random.randn(7, 1)]))
            else:
                gen_birds.append(Bird(self.model_mutate(self.best_bird_weights)))

        return gen_birds

    def save_weights(self, weights_list, filename):
        """
            Save the weights as 1 file per layer weights.
            Not compatible with more than 2 layers.
        """
        np.savetxt(filename+".001", weights_list[0])
        np.savetxt(filename+".002", weights_list[1])

    def load_weights(self, filename):
        """
            Returns loaded weights, already reshaped and ready to use.
            Not compatible with more than 2 layers.
        """
        loaded_weights = []

        loaded_weights.append(np.loadtxt(filename+".001"))
        loaded_weights[0] = loaded_weights[0].reshape(BIRD['WEIGHTS1'][0] , BIRD['WEIGHTS1'][1])

        loaded_weights.append(np.loadtxt(filename+".002"))
        loaded_weights[1] = loaded_weights[1].reshape(BIRD['WEIGHTS2'][0] , BIRD['WEIGHTS2'][1])

        return loaded_weights

if __name__ == "__main__":
    game = FlappBird()
    game.run()
